File: backend-node/routes/cacheRoutes.js
import express from 'express';

const router = express.Router();

// Mock cache settings
let cacheSettings = {
  enabled: true,
  model_aware: true,
  semantic_enabled: false
};

// Mock cache data
const mockCacheEntries = [
  { id: 1, key: 'prompt:123', value: 'cached response', timestamp: Date.now() }
];

// GET /api/cache/stats
router.get('/cache/stats', (req, res) => {
  res.json({
    stats: {
      hits: 10,
      misses: 5,
      total_entries: mockCacheEntries.length,
      hit_rate: 0.67
    }
  });
});

// GET /api/cache/entries
router.get('/cache/entries', (req, res) => {
  const limit = parseInt(req.query.limit) || 100;
  res.json({ entries: mockCacheEntries.slice(0, limit) });
});

// GET /api/cache/enabled
router.get('/cache/enabled', (req, res) => {
  res.json({ enabled: cacheSettings.enabled });
});

// POST /api/cache/enabled
router.post('/cache/enabled', (req, res) => {
  const { enabled } = req.body;
  cacheSettings.enabled = enabled !== undefined ? enabled : true;
  res.json({ enabled: cacheSettings.enabled });
});

// GET /api/cache/model-aware
router.get('/cache/model-aware', (req, res) => {
  res.json({ model_aware: cacheSettings.model_aware });
});

// POST /api/cache/model-aware
router.post('/cache/model-aware', (req, res) => {
  const { model_aware } = req.body;
  cacheSettings.model_aware = model_aware !== undefined ? model_aware : true;
  res.json({ model_aware: cacheSettings.model_aware });
});

// GET /api/cache/semantic-enabled
router.get('/cache/semantic-enabled', (req, res) => {
  res.json({ semantic_enabled: cacheSettings.semantic_enabled });
});

// POST /api/cache/semantic-enabled
router.post('/cache/semantic-enabled', (req, res) => {
  const { semantic_enabled } = req.body;
  cacheSettings.semantic_enabled = semantic_enabled !== undefined ? semantic_enabled : false;
  res.json({ semantic_enabled: cacheSettings.semantic_enabled });
});

// POST /api/cache/clear
router.post('/cache/clear', (req, res) => {
  const deletedCount = mockCacheEntries.length;
  mockCacheEntries.length = 0;
  res.json({ deleted_count: deletedCount });
});

// POST /api/cache/clear-expired
router.post('/cache/clear-expired', (req, res) => {
  // Mock clearing expired entries
  const deletedCount = 2;
  res.json({ deleted_count: deletedCount });
});

export default router;

File: backend-python/routes/cache_routes.py
from flask import Blueprint, request
from services import cache
from utils import handle_errors

cache_bp = Blueprint('cache', __name__)


def create_cache_setting_routes(path, key, getter_fn, setter_fn, default_value=True):
    """
    Factory function to create GET/POST route pairs for cache settings.

    Args:
        path: API path (e.g., '/api/cache/enabled')
        key: JSON key for the setting (e.g., 'enabled')
        getter_fn: Function to get current value
        setter_fn: Function to set new value
        default_value: Default value for POST requests
    """
    # Create unique endpoint names based on the key
    get_endpoint = f'get_{key}'
    set_endpoint = f'set_{key}'

    @cache_bp.route(path, methods=['GET'], endpoint=get_endpoint)
    @handle_errors
    def get_setting():
        return {key: getter_fn()}

    @cache_bp.route(path, methods=['POST'], endpoint=set_endpoint)
    @handle_errors
    def set_setting():
        value = request.json.get(key, default_value)
        setter_fn(value)
        return {key: getter_fn()}

    # Return functions to allow endpoint name customization if needed
    return get_setting, set_setting


@cache_bp.route('/api/cache/stats', methods=['GET'])
@handle_errors
def get_cache_stats():
    """Get cache statistics"""
    return {'stats': cache.get_stats()}


@cache_bp.route('/api/cache/entries', methods=['GET'])
@handle_errors
def get_cache_entries():
    """Get all cache entries"""
    limit = request.args.get('limit', 100, type=int)
    return {'entries': cache.get_all_entries(limit=limit)}


@cache_bp.route('/api/cache/clear', methods=['POST'])
@handle_errors
def clear_cache():
    """Clear all cache entries"""
    return {'deleted_count': cache.clear_all()}


@cache_bp.route('/api/cache/clear-expired', methods=['POST'])
@handle_errors
def clear_expired_cache():
    """Clear expired cache entries"""
    return {'deleted_count': cache.clear_expired()}


# Create GET/POST route pairs using factory
get_cache_enabled, set_cache_enabled = create_cache_setting_routes(
    '/api/cache/enabled',
    'enabled',
    cache.is_enabled,
    cache.set_enabled,
    default_value=True
)

get_cache_model_aware, set_cache_model_aware = create_cache_setting_routes(
    '/api/cache/model-aware',
    'model_aware',
    cache.is_model_aware_cache,
    cache.set_model_aware_cache,
    default_value=True
)

get_semantic_cache_enabled, set_semantic_cache_enabled = create_cache_setting_routes(
    '/api/cache/semantic-enabled',
    'semantic_enabled',
    cache.is_semantic_cache_enabled,
    cache.set_semantic_cache_enabled,
    default_value=False
)
